File: 105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if (inorder.size() == 0) return NULL;
        vector<int>::iterator it;
        int ind = find (inorder.begin(), inorder.end(), preorder[0]) - inorder.begin();
        preorder.erase(preorder.begin());
        TreeNode* root = new TreeNode(inorder[ind]);
        vector <int> left(inorder.begin(), inorder.begin() + ind);
        vector <int> right(inorder.begin() + ind + 1, inorder.end());
        root->left = buildTree(preorder, left);
        root->right = buildTree(preorder, right);
        return root;
    }
};

File: 105. Construct Binary Tree from Preorder and Inorder Traversal.py
# memory exceed

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if len(preorder) == 0:
            return []
        root = TreeNode(preorder[0])
        leftin = inorder[:inorder.index(preorder[0])]
        rightin = inorder[inorder.index(preorder[0])+1:]
        leftpre = preorder[1:1+len(leftin)]
        rightpre = preorder[1+len(leftin):len(preorder)]
        root.left = self.build(leftpre, leftin)
        root.right = self.build(rightpre, rightin)
        return root
        
    def build(self, preorder, inorder):
        if len(preorder) == 0:
            return
        root = TreeNode(preorder[0])
        leftin = inorder[:inorder.index(preorder[0])]
        rightin = inorder[inorder.index(preorder[0])+1:]
        leftpre = preorder[1:1+len(leftin)]
        rightpre = preorder[1+len(leftin):len(preorder)]
        root.left = self.build(leftpre, leftin)
        root.right = self.build(rightpre, rightin)
        return root
